### DISCUSSION TITLE
.add_distance() not working anymore?

### DISCUSSION THREAD

**Discussion by markste-in** [theOehrly/Fast-F1#131] [General]
I used the .add_distance() function a few times before but somehow it is not woking anymore isn't it? Or is it depreciated?

Minimal Code Example:
```
import fastf1 as ff1
ff1.Cache.enable_cache('cache')

event = ff1.get_session(2022, 1, f'FP2')
event.load()
for driver in event.laps.Driver:
    for driver_lap in event.laps.pick_driver(driver).iterlaps():
        try:
            driver_tel = driver_lap[1].get_car_data().add_distance()
        except Exception as e:
            print(f"Exception {e}")
```
Result:
```
core           INFO 	Finished loading data for 20 drivers: ['1', '55', '47', '44', '31', '27', '24', '23', '22', '20', '18', '16', '14', '11', '10', '6', '4', '3', '63', '77']
Exception Telemetry does not contain required channels 'Time' and 'Speed'.
Exception Telemetry does not contain required channels 'Time' and 'Speed'.
Exception Telemetry does not contain required channels 'Time' and 'Speed'.
Exception Telemetry does not contain required channels 'Time' and 'Speed'.
Exception Telemetry does not contain required channels 'Time' and 'Speed'.
Exception Telemetry does not contain required channels 'Time' and 'Speed'.
Exception Telemetry does not contain required channels 'Time' and 'Speed'.
Exception Telemetry does not contain required channels 'Time' and 'Speed'.
Exception Telemetry does not contain required channels 'Time' and 'Speed'.
Exception Telemetry does not contain required channels 'Time' and 'Speed'.
Exception Telemetry does not contain required channels 'Time' and 'Speed'.
Exception Telemetry does not contain required channels 'Time' and 'Speed'.
Exception Telemetry does not contain required channels 'Time' and 'Speed'.
Exception Telemetry does not contain required channels 'Time' and 'Speed'.
Exception Telemetry does not contain required channels 'Time' and 'Speed'.
Exception Telemetry does not contain required channels 'Time' and 'Speed'.
Exception Telemetry does not contain required channels 'Time' and 'Speed'.
Exception Telemetry does not contain required channels 'Time' and 'Speed'.
...
```

**Comment by theOehrly** [comment]
It does work as expected. It's not a problem of add_distance. There are some laps that are missing the necessary timestamps to get the telemetry data.  get_car_data will return an empty Telemetry object and then you'll get the error in add_distance. FastF1 should handle this better or maybe raise a separate exception for this, I guess. 

Something else I noticed, you should not iterate over event.laps.Drivers for this. It's a Series with one entry for each lap! You could use event.laps.Drivers.unique() or just event.drivers

**Reply by markste-in** [reply]
Oh yeah exactly! ... I actually fixed that in my other script already by iterating over event.results.Abbreviation.

I think it would be better if add_distance fills Distance etc. with "nan" if the Telemetry object is empty otherwise you would always need a try/exception block where you have to set those values manually to be able to merge it later on

**Reply by theOehrly** [reply]
Hmm, yes, adding an empty Distance column might actually be nicer. But I can't do that immediately, as it's a breaking change if someone is relying on the error being raised. I'll consider changing it in a future version, though.

**Reply by markste-in** [reply]
You could also implement such a fix in a non-breaking fashion with an optional error parameter.  The default value would keep the functions current behaviour. That would have the advantage that the user can decide whether he wants to ignore empty telemetry data and have no distance or thrown an error

Like that:

```
def add_distance(on_error="raise"):
    try:
        pass # Your normal code

    except Exception as e:
        if on_error.lower() == "ignore":
            self.df["Distance"] = np.nan
            ...
            return ...
        elif on_error.lower()=="raise":
            raise e
        else:
            raise Exception(f"Unknown error handling requested: {on_error}")
```

**Reply by theOehrly** [reply]
Yes, that is another option. Maybe I'll do that, I'll consider it.
